src/dmd/json.d

```diff
@@ -857,17 +891,77 @@ public:
 extern (C++) void json_generate(OutBuffer* buf, Modules* modules)
 {
     scope ToJsonVisitor json = new ToJsonVisitor(buf);
-    json.arrayStart();
-    json.generateCompilerInfo();
-    json.generateBuildInfo();
-    for (size_t i = 0; i < modules.dim; i++)
-    {
-        Module m = (*modules)[i];
-        if (global.params.verbose)
-            message("json gen %s", m.toChars());
-        m.accept(json);
-    }
-    json.generateSemantics();
-    json.arrayEnd();
-    json.removeComma();
+
+    if (global.params.jsonFieldFlags == 0)
+    {
+        // Generate the original format, which is just an array
+        // of modules representing their syntax.
+        json.generateModules(modules);
+        json.removeComma();
+    }
+    else
+    {
+        // Generate the new format which is an object where each
+        // output option is its own field.
+
+        json.objectStart();
+        if (global.params.jsonFieldFlags & JsonFieldFlags.compilerInfo)
+        {
+            json.propertyStart("compilerInfo");
+            json.generateCompilerInfo();
+        }
+        if (global.params.jsonFieldFlags & JsonFieldFlags.buildInfo)
+        {
+            json.propertyStart("buildInfo");
+            json.generateBuildInfo();
+        }
+        if (global.params.jsonFieldFlags & JsonFieldFlags.modules)
+        {
+            json.propertyStart("modules");
+            json.generateModules(modules);
+        }
+        if (global.params.jsonFieldFlags & JsonFieldFlags.semantics)
+        {
+            json.propertyStart("semantics");
+            json.generateSemantics();
+        }
+        json.objectEnd();
+    }
+}
+
+/**
+A string listing the name of each JSON field. Useful for errors messages.
+*/
+enum jsonFieldNames = () {
+    string s;
+    string prefix = "";
+    foreach (enumName; __traits(allMembers, JsonFieldFlags))
+    {
+        if (enumName != "none")
```

Comment:
How about:

```d
foreach (idx, enumName; __traits(allMembers, JsonFieldFlags))
{
    if (idx > 0)
    {
        s ~= ", " ~ enumName;
    }
}
```

Reply:
Hmmm, that produces a different string ", compilerInfo, ..." instead of "compilerInfo, ..."

Reply:
Ah it's too late here. I meant:

```d
if (idx > 0)
    s ~= ", ";
s ~= enumName;
```

(the main point was about not hard-coding that the first `enumName` is named "none")

Reply:
Oh...but now you're hardcoding that the first member should be ignored...is that better than hardcoding that the value named "none" should be ignored?

Reply:
Eh...I suppose I don't have a preference for either...so we'll go with your preference :)